server/src/main/java/io/crate/types/DateType.java

```diff
@@ -71,34 +65,36 @@ public Streamer<Long> streamer() {
 
     @Override
     public Long implicitCast(Object value) throws IllegalArgumentException, ClassCastException {
+        long longVal;
         if (value == null) {
             return null;
-        } else if (value instanceof String) {
-
-            var stringVal = (String) value;
-
+        } else if (value instanceof String stringVal) {
             try {
-                LocalDate dt = LocalDate.parse(stringVal, ISO_FORMATTER);
-                return dt.atStartOfDay(ZoneOffset.UTC).toInstant().toEpochMilli();
+                TemporalAccessor dt = TIMESTAMP_PARSER.parseBest(stringVal, LocalDateTime::from, LocalDate::from);
+                LocalDate localDate = LocalDate.from(dt);
+                return localDate.atStartOfDay(ZoneOffset.UTC).toInstant().toEpochMilli();
             } catch (DateTimeParseException ex) {
                 try {
-                    return Long.parseLong(stringVal);
+                    longVal = Long.parseLong(stringVal);
                 } catch (NumberFormatException e) {
                     throw new ClassCastException("Can't cast '" + value + "' to " + getName());
                 }
             }
-
         } else if (value instanceof Double) {
             // we treat float and double values as seconds with milliseconds as fractions
             // see timestamp documentation
-            return ((Number) (((Double) value) * 1000)).longValue();
+            longVal = ((Number) (((Double) value) * 1000)).longValue();
```

Comment:
Not related to the PR change, but: why do we need to multiply here by 1000 since we are going to divide by 1000 later on? We can remove this multiplication, and the division by 1000 when calculating the epochDay, and add a division by 1000 in the `Number` case(L90), and above(L58) when parsing as long.

Reply:
I think the idea here is to avoid rounding off errors by primitive narrowing conversion through to the float/double -> long downcast. Can be seen in action by e.g.:
```
Float.valueOf(1422294644.581f).longValue() = 1422294656
```
vs.
```
Float.valueOf(1422294644.581f * 1000).longValue() / 1000 = 1422294646
```

Could be solved also using `Math.round` which may be even the safer way but more expensive I guess.
I'd prefer to keep it as it as I don't see lot of improvement by your suggestion but no string opinion.

Reply:
thx for explaining!